ExecuteNonQuery deadlocks while running in UI synchronization context

I have upgraded my npgsql driver to the latest version 5.0.7 (from 4.0.12) and afterwards I encounter the following problem:

We have an application that needs to ensure our database is patched to the latest version of our software.
Therefore it needs to run a script that contains a lot of DDL statements (alter table etc).
This may actually take some time, which is why I gave it a command timeout of 0.
When I use the ExecuteNonQueryAsync method everything works fine and the patch is applied in about 10 seconds.
If I remove the if block from the code below and just call ExecuteNonQuery(); the application freezes and the code never completes.

### Steps to reproduce

                        // this method just does the following:
                        // return new NpgsqlCommand(patchScript, connection as NpgsqlConnection);
                        IDbCommand command = NpgsqlFactory.AllocateNpgsqlCommand(patchScript, connection);
                        command.CommandTimeout = 0;

                        System.Data.Common.DbCommand dbc = command as System.Data.Common.DbCommand;
                        if (dbc != null)
                        {
                            // this works fine, takes a few seconds to complete...
                            System.Threading.Tasks.Task<int> task = dbc.ExecuteNonQueryAsync();
                            while (!task.IsCompleted)
                                System.Threading.Thread.Sleep(100);

                            int n = task.Result;
                        }
                        else
                            // this will freeze the application...
                            command.ExecuteNonQuery();

### The issue

When I break the application the code will stay at the line 177 in NpgsqlReadBuffer.cs forever:

                        var read = async
                            ? await buffer.Underlying.ReadAsync(buffer.Buffer, buffer.FilledBytes, toRead, finalCt)
                            : buffer.Underlying.Read(buffer.Buffer, buffer.FilledBytes, toRead);

There seems to be something very wrong. Although it has already completed in the database backend (if I look in pgadmin I can see the last statement executed on this connection was actually the last statement in my patch script), it is still waiting to read something.

If I change the command timeout in my code to 30 seconds and run it again, it actually hits a debug assertion after the System.IO.IOException for the timeout occurs:

Assertion Failed: Abort=Quit, Retry=Debug, Ignore=Continue

   at Npgsql.NpgsqlConnector.DoCancelRequest(Int32 backendProcessId, Int32 backendSecretKey) in C:\Users\foo\Source\Repos\npgsql\src\Npgsql\NpgsqlConnector.cs:line 1560
   at Npgsql.NpgsqlConnector.PerformPostgresCancellation() in C:\Users\foo\Source\Repos\npgsql\src\Npgsql\NpgsqlConnector.cs:line 1533
   at Npgsql.NpgsqlReadBuffer.<<Ensure>g__EnsureLong|40_0>d.MoveNext() in C:\Users\foo\Source\Repos\npgsql\src\Npgsql\NpgsqlReadBuffer.cs:line 226
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
   at Npgsql.NpgsqlReadBuffer.<Ensure>g__EnsureLong|40_0(NpgsqlReadBuffer buffer, Int32 count, Boolean async, Boolean readingNotifications)
   at Npgsql.NpgsqlReadBuffer.Ensure(Int32 count, Boolean async, Boolean readingNotifications) in C:\Users\foo\Source\Repos\npgsql\src\Npgsql\NpgsqlReadBuffer.cs:line 143
   at Npgsql.NpgsqlConnector.<<ReadMessage>g__ReadMessageLong|194_0>d.MoveNext() in C:\Users\foo\Source\Repos\npgsql\src\Np......

When I click retry I can see that the ReadPosition actually has a value of 1 whereas it should have a value of 0.
I assume that means it did read the result somehow but kept waiting until the timeout occured.

                Debug.Assert(ReadBuffer.ReadPosition == 0);

This is strange, when I use your code it freezes. I only removed the await using because of my environment.
error CS8370: Feature 'using declarations' is not available in C# 7.3. Please use language version 8.0 or greater.

Could it be the .NET framework? We are using Visual Studio 2019 and .NET framework 4.8.
I referenced npgsql.dll and all dependent libraries using the .NET Standard 2.0 binaries from nuget.

--- a/src/Npgsql/MultiplexingConnectorPool.cs
+++ b/src/Npgsql/MultiplexingConnectorPool.cs
@@ -284,7 +284,7 @@ namespace Npgsql
                 Debug.Assert(written, $"Failed to enqueue command to {connector.CommandsInFlightWriter}");
 
                 // Purposefully don't wait for I/O to complete
-                var task = command.Write(connector, async: true);
+                var task = command.Write(connector, async: true, flush: false);
                 stats.NumCommands++;
 
                 switch (task.Status)

--- a/src/Npgsql/NpgsqlCommand.cs
+++ b/src/Npgsql/NpgsqlCommand.cs
@@ -896,27 +896,49 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
         void BeginSend(NpgsqlConnector connector)
             => connector.WriteBuffer.Timeout = TimeSpan.FromSeconds(CommandTimeout);
 
-        void CleanupSend()
+        internal Task<bool> Write(NpgsqlConnector connector, bool async, CancellationToken cancellationToken = default)
         {
-            // ReSharper disable once ConditionIsAlwaysTrueOrFalse
-            if (SynchronizationContext.Current != null)  // Check first because SetSynchronizationContext allocates
-                SynchronizationContext.SetSynchronizationContext(null);
-        }
-
-        internal Task Write(NpgsqlConnector connector, bool async, CancellationToken cancellationToken = default)
-        {
+            // Note that for the methods below we implement both sync and async paths to prevent deadlocks while using a single threaded synchronization context
+            // See #3839
             return (_behavior & CommandBehavior.SchemaOnly) == 0
                 ? WriteExecute(connector, async, cancellationToken)
                 : WriteExecuteSchemaOnly(connector, async, cancellationToken);
 
-            async Task WriteExecute(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
+            Task<bool> WriteExecute(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
             {
                 for (var i = 0; i < _statements.Count; i++)
                 {
                     // The following is only for deadlock avoidance when doing sync I/O (so never in multiplexing)
                     ForceAsyncIfNecessary(ref async, i);
+                    if (async)
+                        return WriteExecuteAsync(connector, i, cancellationToken);
 
-                    var statement = _statements[i];
+                    var executeTask = WriteExecute(connector, i, async: false, cancellationToken);
+                    Debug.Assert(executeTask.IsCompleted);
+                    executeTask.GetAwaiter().GetResult();
+                }
+
+                var syncTask = connector.WriteSync(async: false, cancellationToken);
+                Debug.Assert(syncTask.IsCompleted);
+                syncTask.GetAwaiter().GetResult();
+
+                return Task.FromResult(false);
+
+                async Task<bool> WriteExecuteAsync(NpgsqlConnector connector, int statementIndex, CancellationToken cancellationToken)
+                {
+                    for (var i = statementIndex; i < _statements.Count; i++)
+                    {
+                        await WriteExecute(connector, i, async: true, cancellationToken);
+                    }
+
+                    await connector.WriteSync(async: true, cancellationToken);
+
+                    return true;
+                }
+
+                async Task WriteExecute(NpgsqlConnector connector, int statementIndex, bool async, CancellationToken cancellationToken)
+                {
+                    var statement = _statements[statementIndex];
                     var pStatement = statement.PreparedStatement;
 
                     if (pStatement == null || statement.IsPreparing)
@@ -932,7 +954,7 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
 
                         await connector.WriteBind(
                             statement.InputParameters, string.Empty, statement.StatementName, AllResultTypesAreUnknown,
-                            i == 0 ? UnknownResultTypeList : null,
+                            statementIndex == 0 ? UnknownResultTypeList : null,
                             async, cancellationToken);
 
                         await connector.WriteDescribe(StatementOrPortal.Portal, string.Empty, async, cancellationToken);
@@ -942,7 +964,7 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
                         // The statement is already prepared, only a Bind is needed
                         await connector.WriteBind(
                             statement.InputParameters, string.Empty, statement.StatementName, AllResultTypesAreUnknown,
-                            i == 0 ? UnknownResultTypeList : null,
+                            statementIndex == 0 ? UnknownResultTypeList : null,
                             async, cancellationToken);
                     }
 
@@ -951,30 +973,56 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
                     if (pStatement != null)
                         pStatement.LastUsed = DateTime.UtcNow;
                 }
-
-                await connector.WriteSync(async, cancellationToken);
             }
 
-            async Task WriteExecuteSchemaOnly(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
+            Task<bool> WriteExecuteSchemaOnly(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
             {
                 var wroteSomething = false;
                 for (var i = 0; i < _statements.Count; i++)
                 {
                     ForceAsyncIfNecessary(ref async, i);
+                    if (async)
+                        return WriteExecuteSchemaOnlyAsync(connector, i, wroteSomething, cancellationToken);
+                    
+                    var executeSchemaOnlyTask = WriteExecuteSchemaOnly(connector, i, async: false, cancellationToken);
+                    Debug.Assert(executeSchemaOnlyTask.IsCompleted);
+                    wroteSomething = executeSchemaOnlyTask.GetAwaiter().GetResult();
+                }
 
-                    var statement = _statements[i];
+                if (wroteSomething)
+                {
+                    var syncTask = connector.WriteSync(async: false, cancellationToken);
+                    Debug.Assert(syncTask.IsCompleted);
+                    syncTask.GetAwaiter().GetResult();
+                }
+
+                return Task.FromResult(false);
+
+                async Task<bool> WriteExecuteSchemaOnlyAsync(NpgsqlConnector connector, int statementIndex, bool wroteSomething, CancellationToken cancellationToken)
+                {
+                    for (var i = statementIndex; i < _statements.Count; i++)
+                    {
+                        wroteSomething = await WriteExecuteSchemaOnly(connector, i, async: true, cancellationToken);
+                    }
+
+                    if (wroteSomething)
+                        await connector.WriteSync(async: true, cancellationToken);
+
+                    return true;
+                }
+
+                async Task<bool> WriteExecuteSchemaOnly(NpgsqlConnector connector, int statementIndex, bool async, CancellationToken cancellationToken)
+                {
+                    var statement = _statements[statementIndex];
 
                     if (statement.PreparedStatement?.State == PreparedState.Prepared)
-                        continue;   // Prepared, we already have the RowDescription
+                        return false;   // Prepared, we already have the RowDescription
                     Debug.Assert(statement.PreparedStatement == null);
 
                     await connector.WriteParse(statement.SQL, string.Empty, statement.InputParameters, async, cancellationToken);
                     await connector.WriteDescribe(StatementOrPortal.Statement, statement.StatementName, async, cancellationToken);
-                    wroteSomething = true;
+                    return true;
                 }
-
-                if (wroteSomething)
-                    await connector.WriteSync(async, cancellationToken);
             }
         }
 
@@ -994,8 +1042,6 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
 
             await connector.WriteSync(async, cancellationToken);
             await connector.Flush(async, cancellationToken);
-
-            CleanupSend();
         }
 
         async Task SendPrepare(NpgsqlConnector connector, bool async, CancellationToken cancellationToken = default)
@@ -1025,8 +1071,6 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
 
             await connector.WriteSync(async, cancellationToken);
             await connector.Flush(async, cancellationToken);
-
-            CleanupSend();
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
@@ -1062,8 +1106,6 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
 
             await connector.WriteSync(async, cancellationToken);
             await connector.Flush(async, cancellationToken);
-
-            CleanupSend();
         }
 
         #endregion
@@ -1407,12 +1449,11 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
                 throw;
             }
 
-            async Task NonMultiplexingWriteWrapper(NpgsqlConnector connector, bool async, CancellationToken cancellationToken2)
+            async Task NonMultiplexingWriteWrapper(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
             {
                 BeginSend(connector);
-                await Write(connector, async, cancellationToken2);
-                await connector.Flush(async, cancellationToken2);
-                CleanupSend();
+                async = await Write(connector, async, cancellationToken);
+                await connector.Flush(async, cancellationToken);
             }
         }
 

--- a/src/Npgsql/NpgsqlCommand.cs
+++ b/src/Npgsql/NpgsqlCommand.cs
@@ -896,49 +896,20 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
         void BeginSend(NpgsqlConnector connector)
             => connector.WriteBuffer.Timeout = TimeSpan.FromSeconds(CommandTimeout);
 
-        internal Task<bool> Write(NpgsqlConnector connector, bool async, CancellationToken cancellationToken = default)
+        internal Task Write(NpgsqlConnector connector, bool async, bool flush, CancellationToken cancellationToken = default)
         {
-            // Note that for the methods below we implement both sync and async paths to prevent deadlocks while using a single threaded synchronization context
-            // See #3839
             return (_behavior & CommandBehavior.SchemaOnly) == 0
-                ? WriteExecute(connector, async, cancellationToken)
-                : WriteExecuteSchemaOnly(connector, async, cancellationToken);
+                ? WriteExecute(connector, async, flush, cancellationToken)
+                : WriteExecuteSchemaOnly(connector, async, flush, cancellationToken);
 
-            Task<bool> WriteExecute(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
+            async Task WriteExecute(NpgsqlConnector connector, bool async, bool flush, CancellationToken cancellationToken)
             {
                 for (var i = 0; i < _statements.Count; i++)
                 {
                     // The following is only for deadlock avoidance when doing sync I/O (so never in multiplexing)
                     ForceAsyncIfNecessary(ref async, i);
-                    if (async)
-                        return WriteExecuteAsync(connector, i, cancellationToken);
 
-                    var executeTask = WriteExecute(connector, i, async: false, cancellationToken);
-                    Debug.Assert(executeTask.IsCompleted);
-                    executeTask.GetAwaiter().GetResult();
-                }
-
-                var syncTask = connector.WriteSync(async: false, cancellationToken);
-                Debug.Assert(syncTask.IsCompleted);
-                syncTask.GetAwaiter().GetResult();
-
-                return Task.FromResult(false);
-
-                async Task<bool> WriteExecuteAsync(NpgsqlConnector connector, int statementIndex, CancellationToken cancellationToken)
-                {
-                    for (var i = statementIndex; i < _statements.Count; i++)
-                    {
-                        await WriteExecute(connector, i, async: true, cancellationToken);
-                    }
-
-                    await connector.WriteSync(async: true, cancellationToken);
-
-                    return true;
-                }
-
-                async Task WriteExecute(NpgsqlConnector connector, int statementIndex, bool async, CancellationToken cancellationToken)
-                {
-                    var statement = _statements[statementIndex];
+                    var statement = _statements[i];
                     var pStatement = statement.PreparedStatement;
 
                     if (pStatement == null || statement.IsPreparing)
@@ -954,7 +925,7 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
 
                         await connector.WriteBind(
                             statement.InputParameters, string.Empty, statement.StatementName, AllResultTypesAreUnknown,
-                            statementIndex == 0 ? UnknownResultTypeList : null,
+                            i == 0 ? UnknownResultTypeList : null,
                             async, cancellationToken);
 
                         await connector.WriteDescribe(StatementOrPortal.Portal, string.Empty, async, cancellationToken);
@@ -964,7 +935,7 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
                         // The statement is already prepared, only a Bind is needed
                         await connector.WriteBind(
                             statement.InputParameters, string.Empty, statement.StatementName, AllResultTypesAreUnknown,
-                            statementIndex == 0 ? UnknownResultTypeList : null,
+                            i == 0 ? UnknownResultTypeList : null,
                             async, cancellationToken);
                     }
 
@@ -973,56 +944,36 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
                     if (pStatement != null)
                         pStatement.LastUsed = DateTime.UtcNow;
                 }
+
+                await connector.WriteSync(async, cancellationToken);
+
+                if (flush)
+                    await connector.Flush(async, cancellationToken);
             }
 
-            Task<bool> WriteExecuteSchemaOnly(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
+            async Task WriteExecuteSchemaOnly(NpgsqlConnector connector, bool async, bool flush, CancellationToken cancellationToken)
             {
                 var wroteSomething = false;
                 for (var i = 0; i < _statements.Count; i++)
                 {
                     ForceAsyncIfNecessary(ref async, i);
-                    if (async)
-                        return WriteExecuteSchemaOnlyAsync(connector, i, wroteSomething, cancellationToken);
-                    
-                    var executeSchemaOnlyTask = WriteExecuteSchemaOnly(connector, i, async: false, cancellationToken);
-                    Debug.Assert(executeSchemaOnlyTask.IsCompleted);
-                    wroteSomething = executeSchemaOnlyTask.GetAwaiter().GetResult();
-                }
 
-                if (wroteSomething)
-                {
-                    var syncTask = connector.WriteSync(async: false, cancellationToken);
-                    Debug.Assert(syncTask.IsCompleted);
-                    syncTask.GetAwaiter().GetResult();
-                }
-
-                return Task.FromResult(false);
-
-                async Task<bool> WriteExecuteSchemaOnlyAsync(NpgsqlConnector connector, int statementIndex, bool wroteSomething, CancellationToken cancellationToken)
-                {
-                    for (var i = statementIndex; i < _statements.Count; i++)
-                    {
-                        wroteSomething = await WriteExecuteSchemaOnly(connector, i, async: true, cancellationToken);
-                    }
-
-                    if (wroteSomething)
-                        await connector.WriteSync(async: true, cancellationToken);
-
-                    return true;
-                }
-
-                async Task<bool> WriteExecuteSchemaOnly(NpgsqlConnector connector, int statementIndex, bool async, CancellationToken cancellationToken)
-                {
-                    var statement = _statements[statementIndex];
+                    var statement = _statements[i];
 
                     if (statement.PreparedStatement?.State == PreparedState.Prepared)
-                        return false;   // Prepared, we already have the RowDescription
+                        continue;   // Prepared, we already have the RowDescription
                     Debug.Assert(statement.PreparedStatement == null);
 
                     await connector.WriteParse(statement.SQL, string.Empty, statement.InputParameters, async, cancellationToken);
                     await connector.WriteDescribe(StatementOrPortal.Statement, statement.StatementName, async, cancellationToken);
-                    return true;
+                    wroteSomething = true;
                 }
+
+                if (wroteSomething)
+                    await connector.WriteSync(async, cancellationToken);
+
+                if (flush)
+                    await connector.Flush(async, cancellationToken);
             }
         }
 
@@ -1449,11 +1400,10 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
                 throw;
             }
 
-            async Task NonMultiplexingWriteWrapper(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
+            Task NonMultiplexingWriteWrapper(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
             {
                 BeginSend(connector);
-                async = await Write(connector, async, cancellationToken);
-                await connector.Flush(async, cancellationToken);
+                return Write(connector, async, flush: true, cancellationToken);
             }
         }
 

--- a/src/Npgsql/NpgsqlCommand.cs
+++ b/src/Npgsql/NpgsqlCommand.cs
@@ -970,10 +970,11 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
                 }
 
                 if (wroteSomething)
+                {
                     await connector.WriteSync(async, cancellationToken);
-
-                if (flush)
-                    await connector.Flush(async, cancellationToken);
+                    if (flush)
+                        await connector.Flush(async, cancellationToken);
+                }
             }
         }
 
@@ -1316,7 +1317,8 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
                         // Instead, all sends for non-first statements are performed asynchronously (even if the user requested sync),
                         // in a special synchronization context to prevents a dependency on the thread pool (which would also trigger
                         // deadlocks).
-                        sendTask = NonMultiplexingWriteWrapper(connector, async, CancellationToken.None);
+                        BeginSend(connector);
+                        sendTask = Write(connector, async, flush: true, CancellationToken.None);
 
                         // The following is a hack. It raises an exception if one was thrown in the first phases
                         // of the send (i.e. in parts of the send that executed synchronously). Exceptions may
@@ -1398,12 +1400,6 @@ GROUP BY pg_proc.proargnames, pg_proc.proargtypes, pg_proc.proallargtypes, pg_pr
                 }
                     
                 throw;
-            }
-
-            Task NonMultiplexingWriteWrapper(NpgsqlConnector connector, bool async, CancellationToken cancellationToken)
-            {
-                BeginSend(connector);
-                return Write(connector, async, flush: true, cancellationToken);
             }
         }
 

--- a/test/Npgsql.Tests/BugTests.cs
+++ b/test/Npgsql.Tests/BugTests.cs
@@ -1322,5 +1322,26 @@ $$;");
                 Assert.AreEqual(2, await exporter.ReadAsync<int?>());
             }
         }
+
+        [Test, Timeout(5000)]
+        [IssueLink("https://github.com/npgsql/npgsql/issues/3839")]
+        public async Task SingleTreadedSynchronizationContextDeadlock()
+        {
+            var syncContext = new SingleThreadSynchronizationContext(nameof(SingleTreadedSynchronizationContextDeadlock));
+            using (var _ = syncContext.Enter())
+            {
+                // We have to Yield, so the current thread is changed to the one used by SingleThreadSynchronizationContext
+                await Task.Yield();
+                using var connection = OpenConnection();
+
+                var data = new string('x', 5_000_000);
+                using var cmd = new NpgsqlCommand("SELECT generate_series(1, 500000); SELECT @p", connection);
+                cmd.Parameters.AddWithValue("p", NpgsqlDbType.Text, data);
+                cmd.ExecuteNonQuery();
+            }
+
+            // We have to make another Yield to change the current thread from the one used by SingleThreadSynchronizationContext
+            await Task.Yield();
+        }
     }
 }

--- a/test/Npgsql.Tests/BugTests.cs
+++ b/test/Npgsql.Tests/BugTests.cs
@@ -1325,9 +1325,9 @@ $$;");
 
         [Test, Timeout(5000)]
         [IssueLink("https://github.com/npgsql/npgsql/issues/3839")]
-        public async Task SingleTreadedSynchronizationContextDeadlock()
+        public async Task SingleThreadedSynchronizationContextDeadlock()
         {
-            var syncContext = new SingleThreadSynchronizationContext(nameof(SingleTreadedSynchronizationContextDeadlock));
+            var syncContext = new SingleThreadSynchronizationContext(nameof(SingleThreadedSynchronizationContextDeadlock));
             using (var _ = syncContext.Enter())
             {
                 // We have to Yield, so the current thread is changed to the one used by SingleThreadSynchronizationContext

